File: src/z1_walkingpad_mcp/metrics.py
"""Local health-metric estimation — the Z1 streams speed/distance/time/steps
but NOT calories, so we compute energy expenditure with the **ACSM walking
metabolic equation** (level grade), the exercise-physiology standard:

    VO2 (ml/kg/min) = 0.1 * speed(m/min) + 3.5          (grade = 0)
    kcal/min        = VO2 * weight_kg / 200             (5 kcal per L O2)

Best validated for 50-100 m/min (~3-6 km/h); below that (our minimum is
1.6 km/h) expect somewhat larger error. Classic validation puts the standard
error around 2.0-2.6 ml/kg/min (~0.6-0.7 MET); a 2021 field comparison found
~13% overprediction for unloaded walking. This replaces the coarser
Compendium-of-METs bucket table, which research shows misclassifies
intensity near slow-walking speeds.

Sources: ACSM Guidelines for Exercise Testing and Prescription;
japplphysiol.00121.2021 (2021); pubmed 16095415; pubmed 35876127 (2022).
"""

from __future__ import annotations

import os

DEFAULT_WEIGHT_KG = 75.0

# resting component of the ACSM equation (3.5 ml/kg/min = 1 MET)
_RESTING_VO2 = 3.5
# walking economy: 0.1 ml/kg/min per m/min on level ground
_SPEED_COEFF = 0.1


def configured_weight_kg() -> float:
    raw = os.environ.get("Z1_WEIGHT_KG")
    if raw:
        try:
            return float(raw)
        except ValueError:
            pass
    return DEFAULT_WEIGHT_KG


def vo2_for_speed(kmh: float) -> float:
    """Gross VO2 (ml/kg/min) per the ACSM level-walking equation."""
    speed_m_per_min = max(0.0, kmh) * 1000 / 60
    return _SPEED_COEFF * speed_m_per_min + _RESTING_VO2


def met_for_speed(kmh: float) -> float:
    return vo2_for_speed(kmh) / _RESTING_VO2


def kcal_per_minute(kmh: float, weight_kg: float) -> float:
    return vo2_for_speed(kmh) * weight_kg / 200


class CalorieTracker:
    """Integrates calorie burn from a stream of speed samples."""

    def __init__(self, weight_kg: float | None = None) -> None:
        self.weight_kg = weight_kg if weight_kg is not None else configured_weight_kg()
        self.total_kcal = 0.0

    def add_sample(self, speed_kmh: float, elapsed_s: float) -> None:
        """Credit burn for elapsed_s spent at speed_kmh."""
        if elapsed_s <= 0:
            return
        self.total_kcal += kcal_per_minute(speed_kmh, self.weight_kg) * (elapsed_s / 60)

    def reset(self) -> None:
        self.total_kcal = 0.0

File: macos/Sources/Z1Core/Metrics.swift
import Foundation

/// Local health-metric estimation — the Z1 streams speed/distance/time/steps
/// but NOT calories, so we compute energy expenditure the same way fitness apps
/// do: the Compendium of Physical Activities MET values for level walking.
///
///     kcal/min = MET * 3.5 * weight_kg / 200
///
/// Mirrors `metrics.py`.
public enum Z1Metrics {
    /// MET by walking speed (km/h), level surface. Linear interpolation between points.
    public static let metTable: [(speed: Double, met: Double)] = [
        (0.0, 1.0), // standing
        (1.6, 2.0), // very slow walk
        (2.5, 2.8),
        (3.2, 3.0),
        (4.0, 3.5),
        (4.8, 3.8),
        (5.5, 4.3),
        (6.4, 5.0), // Z1 max speed
    ]

    public static let defaultWeightKg = 75.0

    public static func metForSpeed(_ kmh: Double) -> Double {
        if kmh <= metTable[0].speed { return metTable[0].met }
        for i in 1 ..< metTable.count {
            let (s0, m0) = metTable[i - 1]
            let (s1, m1) = metTable[i]
            if kmh <= s1 {
                let frac = (kmh - s0) / (s1 - s0)
                return m0 + frac * (m1 - m0)
            }
        }
        return metTable[metTable.count - 1].met
    }

    public static func kcalPerMinute(_ kmh: Double, weightKg: Double) -> Double {
        metForSpeed(kmh) * 3.5 * weightKg / 200
    }
}

/// Integrates calorie burn from a stream of speed samples.
public struct CalorieTracker: Sendable {
    public var weightKg: Double
    public private(set) var totalKcal = 0.0

    public init(weightKg: Double = Z1Metrics.defaultWeightKg) {
        self.weightKg = weightKg
    }

    /// Credit burn for `elapsedS` seconds spent at `speedKmh`.
    public mutating func addSample(speedKmh: Double, elapsedS: Double) {
        guard elapsedS > 0 else { return }
        totalKcal += Z1Metrics.kcalPerMinute(speedKmh, weightKg: weightKg) * (elapsedS / 60)
    }

    public mutating func reset() {
        totalKcal = 0.0
    }
}
